central/signatureintegration/datastore/datastore_impl.go

```diff
@@ -149,6 +163,67 @@ func (d *datastoreImpl) verifyIntegrationIDDoesNotExist(ctx context.Context, id
 	return nil
 }
 
+func (d *datastoreImpl) verifyIntegrationIDIsNotInPolicy(ctx context.Context, id string) error {
+	var useAllAccessCtx bool
+	if err := sac.VerifyAuthzOK(policySAC.ReadAllowed(ctx)); err != nil {
+		useAllAccessCtx = true
+	}
+
+	// Only create a context with all access if we are not allowed to read policies.
+	policyCtx := ctx
+	if useAllAccessCtx {
+		policyCtx = sac.WithAllAccess(ctx)
+	}
+
+	policies, err := d.policyStore.GetAllPolicies(policyCtx)
+	if err != nil {
+		return errors.Wrap(err, "retrieving all policies")
+	}
+
+	var policiesContainingID []string
+	for _, p := range policies {
+		if checkIfPolicyContainsID(id, p) {
+			policiesContainingID = append(policiesContainingID, p.GetName())
+		}
+	}
+
+	if len(policiesContainingID) != 0 {
+		integration, err := d.getSignatureIntegrationByID(ctx, id)
+		if err != nil {
+			return err
+		}
+
+		errMsg := fmt.Sprintf("cannot delete signature integration %q since there are existing policies that"+
+			"reference it", integration.GetName())
+
+		// Only return list of policies with references to signature integration if the context had read access to
+		// policies. Otherwise, we would potentially leak names.
+		if !useAllAccessCtx {
```

Comment:
I've added the intersection of the policies visible to the user with the policies that have references, policies which are not visible to the user will be shown as "<hidden>" within the error message.
I chose to not repeat the "<hidden>" per policy that is not visible to the user, but rather only add one, otherwise the error message could look like this:
`.. existing policies that reference it: [policyA, policyB, <hidden>, <hidden>]`